Avoid a situation where no NVDA modifier key is defined

CyrilleB79 opened the pull request:
### Link to issue number:

Fixes #14527

May be considered a follow-up of #14233.

### Summary of the issue:
When configuring NVDA Modifier Key for a non-default profile and for the default profile, we can end-up with no NVDA key defined at all in the non-default profile (see #14527 for detailed steps).

This is due to the fact that the config stores one item per NVDA key but that there is a dependency between these values to honor the requirement that at least one key should be defined as NVDA key. These values are checked when changing NVDA keys via the GUI. But the requirement may be missed when switching profiles and stacking their configurations.

There is also a second scenario allowing to have no NVDA key defined using only default profile:
* In keyboard settings enable only caps lock and disable both insert, then validate
* Open the NVDA startup dialog via Help menu, uncheck caps lock and validate.
This scenario is quite unlikely but there is no check in NVDA startup dialog to prevent such situation.
* 

### Description of user facing changes
Situations where no NVDA key is defined will not be possible anymore.

### Description of development approach
* Use only one config item to store the used NVDA modifier keys.
* Define a config flag for NVDA key; each value corresponds to one bit so that all the values can be bitwise or-ed and the reesulting value can be stored in the config as an integer.

During the dev, I have had to import key labels (via the config flag) early during NVDA startup. This has caused issues since `pgettext` was not yet defined at this moment. I have tried to fix this; but please double-check it because I do not know very well this part of the code and why gettext translations need to be initialized two times during NVDA startup.

Note: with this PR, `keyboardHandler.SUPPORTED_NVDA_MODIFIER_KEY` is not used anymore in NVDA's code. Should I deprecate it?

### Testing strategy:

Manual tests:
* Check config upgrade in `nvda.ini`
* Modify NVDA modifier keys in keyboard parameters
* Enable or not caps lock as NVDA modifier in NVDA startup dialog
* Test scenario of #14527
* Test 2nd scenario with startup dialog described in this PR.

Unit tests:
* New unit tests created to check various config upgrade situations

### Known issues with pull request:
None

### Change log entries:

Bug fixes
When configuring NVDA it will be ensured that at least one key is defined as NVDA key in any situation. (#14527)

Deprecation:
keyboardHandler.SUPPORTED_NVDA_MODIFIER_KEYS is deprecated with no direct replacement. Consider using the class config.configFlags.NVDAKey instead. (#14528)

API breaking changes:
- In the configuration, the following keys have been removed from the ``[keyboard]`` section: useCapsLockAsNVDAModifierKey, useNumpadInsertAsNVDAModifierKey, useExtendedInsertAsNVDAModifierKey. They are replaced by NVDAModifierKeys. (#14528)
  - 
### Code Review Checklist:


- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

CyrilleB79 committed:
Avoid a situation where no NVDA modifier key is defined

CyrilleB79 committed:
Fix bug when the user tries to configure no key as NVDA modifier key through startup dialog and remove unneeded imports.

CyrilleB79 committed:
Trial to fix install issue

CyrilleB79 committed:
Small fix in log message.

seanbudd commented on source/gui/startupDialogs.py:
```diff
@@ -104,7 +105,26 @@ def __init__(self, parent):
+			gui.messageBox(
+				_(
+					# Translators: The title of an error message box displayed when validating the startup dialog
+					"Despite your choice, caps lock will remain used as NVDA modifier key"
+					" since no other possible key is currently configured as NVDA modifier key."
```
I've suggested a rephrase of this error message to be more clear what the choice was, and why it failed

```suggestion
					"No NVDA modifier key has been selected. "
					"As an NVDA modifier key must be set, caps lock will be saved as the NVDA modifier key. "
```

CyrilleB79 replied:
I have opted for another phrasing in 61af5f252be7acb40e51cb7aef5f0992a90b3e79 since I do not think that NVDAKey configuration should be modified in this case.

seanbudd commented on source/gui/startupDialogs.py:
```diff
@@ -104,7 +105,26 @@ def __init__(self, parent):
+				_("Error"),
+				wx.OK | wx.ICON_ERROR,
+				self,
+			)
```
I think this should be explicitly set here, in case the user had a different initial config.
Otherwise the message should indicate that the NVDA modifier key will not change, rather than be set to caps lock.

```suggestion
			)
			config.conf["keyboard"]["NVDAModifierKeys"] = NVDAKeysVal.CAPS_LOCK
```

CyrilleB79 replied:
When there is an issue with the checkbox value, it's better not to save anything new in the config regarding NVDA key. In 61af5f252be7acb40e51cb7aef5f0992a90b3e79, the messagenow  indicates that nothing will change.

seanbudd commented on source/nvda.pyw:
```diff
@@ -63,13 +65,18 @@ import locale
 except:
 	gettext.install('nvda')
+	# Install our pgettext function.
+	builtins.pgettext = lambda ctx, msg: msg
```
This should be consistent with the other instance in the unit tests
It might be worth turning this into a named function e.g. `def pgettextNoTranslation`

```suggestion
	builtins.pgettext = lambda context, message: message
```

CyrilleB79 replied:
I have added your suggestion in 61af5f252be7acb40e51cb7aef5f0992a90b3e79. I have kept this function anonymous. Defining a named function here would make the code more heavy IMO. And I did not want either to add an extra import for this function (its always not very clear when things need to be imported in nvda.ppw without risking to import things too early).

CyrilleB79 replied:
Also, should I add the same `pgettext` function in `nvda_slave.pyw`? I have not found any case where `pgettext` missing causes a problem there; but maybe its better to handle `nvda.pyw` and `nvda_slave.pyw` the same way.

What do you think?

seanbudd replied:
Probably best to add it to `nvda_slave` just in case

CyrilleB79 replied:
Done in 214c4e4efb767fd6110620c90d3c02000627fd9d.

CyrilleB79 committed:
Address review comments.

seanbudd commented on source/gui/startupDialogs.py:
```diff
@@ -115,8 +115,8 @@ def onOk(self, evt):
 					# Translators: The title of an error message box displayed when validating the startup dialog
-					"Despite your choice, caps lock will remain used as NVDA modifier key"
-					" since no other possible key is currently configured as NVDA modifier key."
+					"At least one NVDA modifier key should be selected. "
+					"As a consequence it has not been possible to disable the caps lock key."
```
good catch, edited my suggestion

CyrilleB79 committed:
Update source/gui/startupDialogs.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

CyrilleB79 committed:
Added pgettext also to nvda_slave

CyrilleB79 commented:
@seanbudd I have addressed all the review comments. 

However I have raised a new question while addressing the review. Please answer before merging. Thanks.

seanbudd commented:
>Note: with this PR, keyboardHandler.SUPPORTED_NVDA_MODIFIER_KEY is not used anymore in NVDA's code. Should I deprecate it?

Yes, this is a good idea. Suggest `NVDAKey` as a replacement

seanbudd reviewed (commented):
Let me know if I missed any other questions

CyrilleB79 commented:
> Let me know if I missed any other questions

Yes please, your feedback to this comment in `nvda.pyw`:

> While working on the review in `nvda_slave.pyw`, I realize that the `install` method does not take any parameter there, wherease here in `nvda.pyw`, it takes `True` as parameter. This was already the case before this PR.
> 
> The help on this `install` method is as follows:
> ```
> >>> help(trans.install)
> Help on method install in module gettext:
> 
> install(names=None) method of gettext.GNUTranslations instance
> 
> ```
> 
> 
> And the documentation indicates:
> 
> > install(names=None)
> > This method installs gettext() into the built-in namespace, binding it to _.
> > If the names parameter is given, it must be a sequence containing the names of functions you want to install in the builtins namespace in addition to _(). Supported names are 'gettext', 'ngettext', 'lgettext' and 'lngettext'.
> > 
> 
> This `True` parameter does not make sens to me. Is it a bug or something left over after a change in the signature of the `install` method?
> 
> Should I remove this parameter? If yes, should I do it here or in a separate PR (in case this causes unexpected issue and need to be reverted)?
> 
> 
>

seanbudd commented:
I agree, remove the parameter in this PR.

CyrilleB79 committed:
Remove inappropriate True parameter for pgettext install function.

CyrilleB79 committed:
Deprecate SUPPORTED_NVDA_MODIFIER_KEYS

CyrilleB79 commented:
I have addressed the parameter issue and implemented the deprecation. I have also updated the initial description's change log with the deprecation info.

The PR is ready to review again.

seanbudd reviewed (approved):
Thanks, LGTM

seanbudd committed:
update changes

seanbudd committed:
Merge remote-tracking branch 'origin/master' into nvdaKey
